lib/active_merchant/billing/gateways/worldpay.rb

```diff
@@ -457,7 +457,12 @@ def default_address
       end
 
       def parse(action, xml)
-        doc = Nokogiri::XML(xml)
+        xml = xml.strip.gsub(/\&/, '&amp;')
+        begin
+          doc = Nokogiri::XML(xml, &:strict)
+        rescue Nokogiri::XML::SyntaxError => e
+          raise InvalidResponseError, "There was an error parsing the response: #{e} | The original response body was: #{xml}"
```

Comment:
Honestly this feels a little weird; is this pattern of `raise`ing within a `rescue` bad? Acceptable? Good? Is there a more idiomatic way to convert one type of error into another?

Reply:
Also, `InvalidResponseError` is only used one other place in AM, for catching a certain type of error thrown for any gateway connection: https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/network_connection_retries.rb#L31-L32

Reply:
Exceptions as control flow is typically a "no no", but often folks take this approach. I guess the thing to weigh here is:

1) is it hard to not use an exception as control flow
* if so, why?
* if not, another pattern should probably be used.
2) are exceptions as control flow already present in the system, if so it might be worth continuing the pattern.

Reply:
@jeremywrowe Hmm, ok. The answer to 1. is not particularly obvious to me (nor is 2. really). 

My initial thinking was, we have a library (Nokogiri) in this library (AM) that throws an exception in certain  apparently-not-rare circumstances. It seems better to (A) wrap it in an abstracted AM-level exception rather than (B) having the original bubble all the way to the top. Would B be the alternative here, or is there an option I'm not seeing (very likely)? Or is B actually favorable?

Reply:
If we need to raise an exception it's probably best to wrap it and include the original context. This changes depending on if the code being described is a library or not. Since this is library code, it's probably best to have a single exception to catch rather than catching everything.

The favorable thing here is to not throw an exception at all, but instead, mark the response as an error and provide a reason why. I think this boils down to making a judgment call though - Since I haven't taken a look at all of what it might take to do so.

The paths that (in my opinion) are ranked below:

1) Respond with a failure response with no exception
2) raise a library-specific exception with the original exception/context included
3) raise the original exception

Reply:
Thanks, that helps. I moved to this approach primarily because Spreedly would already be treating the outcome  appropriately, but we should probably accommodate the change there rather than take an approach which is suboptimal for AM. There are some problems remaining with the original approach however. I'm going to move discussion back there https://github.com/activemerchant/active_merchant/pull/3380